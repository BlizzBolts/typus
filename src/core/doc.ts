import ts from "typescript";

export interface SerialziedDoc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: SerialziedDoc[];
  parameters?: SerialziedDoc[];
  children?: SerialziedDoc[];
  symbol?: ts.Symbol;
}
export class Doc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;

  constructor(param: Doc = {} as Doc) {
    this.name = param.name || "";
    this.documentation = param.documentation || "";
    this.type = param.type || "";
    this.tags = param.tags || [];
    this.required = param.required || false;
    this.members = param.members || [];
    this.parameters = param.parameters || [];
    this.children = param.children || [];
    this.symbol = param.symbol || (null as unknown as ts.Symbol);
  }

  static serialize = (curr: Doc): SerialziedDoc => {
    return {
      name: curr.name,
      documentation: curr.documentation,
      type: curr.type,
      tags: curr.tags,
      required: curr.required,
      members: curr.members?.map(Doc.serialize),
      parameters: curr.parameters?.map(Doc.serialize),
      children: curr.children?.map(Doc.serialize),
    };
  };
}
