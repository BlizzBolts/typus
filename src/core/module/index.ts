export { ModuleGraph } from "./graph";
export { Module } from "./module";
export type { BaseModule } from "./types";
