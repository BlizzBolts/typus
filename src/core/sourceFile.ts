import { Doc, SerialziedDoc } from "./doc";

export type FilePath = string;

interface SourceFilePayload {
  filePath: FilePath;
  docs?: Doc[];
}

export interface SerialzieSourceFile {
  filePath: FilePath;
  docs: SerialziedDoc[];
}

export class SourceFile {
  private filePath: FilePath;

  private docs?: Doc[];

  constructor(payload: SourceFilePayload) {
    const { filePath, docs = [] } = payload;
    this.filePath = filePath;
    this.docs = docs;
  }

  public setFilePath(filePath: FilePath): this {
    this.filePath = filePath;
    return this;
  }

  public getFilePath(): FilePath {
    return this.filePath;
  }

  public addDoc(doc: Doc) {
    this.docs!.push(doc);
    return this;
  }

  public static serialize(curr: SourceFile) {
    return {
      filePath: curr.filePath,
      docs: curr.docs?.map(Doc.serialize),
    };
  }
}
