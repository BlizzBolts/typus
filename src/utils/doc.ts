import ts from "typescript";
import { Doc } from "../core/doc";

// export const getInterfaceDocuments = (
//   node: ts.Node,
//   checker: ts.TypeChecker,
// ) => {
//   const symbol = checker.getSymbolAtLocation(node.name)
//   if (!symbol) {
//     return
//   }

//   const main = generateDoc(symbol, checker, node)

//   main.fields = node.members
//     .map((m) => {
//       if (!m.name) return
//       const symbol = checker.getSymbolAtLocation(m.name)
//       if (!symbol) return
//       return generateDoc(symbol, checker, m)
//     })
//     .filter(Boolean) as DocEntry[]

//   return main
// }

export const generateDoc = (
  node: ts.Node,
  type: ts.Type,
  checker: ts.TypeChecker
): Doc => {
  const symbol = checker.getSymbolAtLocation(node) || type.getSymbol();
  return new Doc({
    name: symbol?.getName(),
    documentation: ts.displayPartsToString(
      symbol?.getDocumentationComment(checker)
    ),
    type: checker.typeToString(
      checker.getTypeOfSymbolAtLocation(symbol!, symbol?.valueDeclaration!)
    ),
    tags: symbol?.getJsDocTags(checker).map((o) => {
      return {
        name: o.name,
        text: ts.displayPartsToString(o.text),
      };
    }),
    children: [],
    members: [],
    parameters: [],
    symbol: checker.getTypeAtLocation(node) as unknown as ts.Symbol,
    // required: symbol?.["questionToken"],
  });
};
