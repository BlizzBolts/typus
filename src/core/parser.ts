import ts from "typescript";
import { addCache, getCache, hasCache } from "../cache";
import { serializeSymbol } from "../utils/doc";
import { Doc } from "./doc";

export const defaultCompilerOptions = {
  target: ts.ScriptTarget.ES5,
  module: ts.ModuleKind.CommonJS,
};

export class Parser {
  private compilerOptions: ts.CompilerOptions;
  private filePaths: string | string[];
  private program: ts.Program;
  private typeChecker: ts.TypeChecker;

  constructor(
    filePaths: string | string[],
    compilerOptions: Partial<ts.CompilerOptions>
  ) {
    this.compilerOptions = Object.assign(
      defaultCompilerOptions,
      compilerOptions
    );
    this.filePaths = Array.isArray(filePaths) ? filePaths : [filePaths];
    this.program = ts.createProgram(this.filePaths, this.compilerOptions);
    this.typeChecker = this.program.getTypeChecker();
  }

  parse() {
    const validSourceFiles = this.program.getSourceFiles().filter((o) => {
      return !o.isDeclarationFile;
    });

    const rootDoc = new Doc();

    validSourceFiles.forEach((o) => {
      o.forEachChild((p) => {
        this.traverse(p, rootDoc);
      });
    });

    return rootDoc;
  }

  parseInterface(node: ts.InterfaceDeclaration, parent: Doc) {
    const interfaceType = this.typeChecker.getTypeAtLocation(node.name);
    const Doc = serializeSymbol(interfaceType.getSymbol()!, this.typeChecker);
    parent.children?.push(Doc);
    addCache(interfaceType, Doc);

    if (Array.isArray(node.members)) {
      node.members.forEach((memberNode) => {
        this.traverse(memberNode, Doc);
      });
    }
  }

  parseFunction(node: ts.FunctionDeclaration, parent: Doc) {
    const fnType = this.typeChecker.getTypeAtLocation(node);
    const Doc = serializeSymbol(fnType.getSymbol()!, this.typeChecker);
    parent.children?.push(Doc);
    addCache(fnType, Doc);

    if (Array.isArray(node.parameters)) {
      node.parameters.forEach((parameter) => this.traverse(parameter, Doc));
    }
  }

  parseFunctionParameter = (node: ts.ParameterDeclaration, parent: Doc) => {
    const parameterType = this.typeChecker.getTypeAtLocation(node.name);
    if (hasCache(parameterType)) {
      parent.parameters?.push(getCache(parameterType)!);
    } else {
      const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
      parent.parameters?.push(serializeSymbol(symbol!, this.typeChecker));
    }
  };

  parseInterfaceMember = (node: ts.TypeElement, parent: Doc) => {
    const memberType = this.typeChecker.getTypeAtLocation(node!.name!);

    if (hasCache(memberType)) {
      parent.members?.push(getCache(memberType)!);
    } else {
      const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
      parent.members?.push(serializeSymbol(symbol!, this.typeChecker));
    }
  };

  traverse(node: ts.Node, parent: Doc) {
    if (ts.isInterfaceDeclaration(node)) {
      this.parseInterface(node, parent);
    }

    if (ts.isFunctionDeclaration(node)) {
      this.parseFunction(node, parent);
    }

    if (ts.isParameter(node)) {
      this.parseFunctionParameter(node, parent);
    }

    if (ts.isTypeElement(node)) {
      this.parseInterfaceMember(node, parent);
    }

    if (ts.isVariableDeclaration(node)) {
      console.log("variable!");
    }
  }

  getProgram() {
    return this.program;
  }

  getTypeChecker() {
    return this.typeChecker;
  }
}
