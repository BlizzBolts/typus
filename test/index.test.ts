import { it, expect, suite, beforeEach, afterEach } from "vitest";
import path from "path";
import { Parser } from "../src/core/parser";
import { Doc } from "../src/core/doc";

let parser: Parser = new Parser();

suite("individual parse", () => {
  it("should parse interface", () => {
    parser.setup(path.resolve(__dirname, "./manifest/interface.ts"));
    const docCache = parser.parse();
    expect(Doc.serialize(docCache)).toMatchSnapshot();
  });

  it("should parse interface with external deps", () => {});

  it("should parse class", () => {});

  it("should parse class with external deps", () => {});

  it("should parse variables", () => {});

  it("should parse type", () => {});

  it("should parse function", () => {});

  it("should parse function with external deps", () => {});
});
