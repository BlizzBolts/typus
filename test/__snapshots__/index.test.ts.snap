// Vitest Snapshot v1

exports[`individual parse > should parse interface 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "documentation": "Returns the average of two numbers.",
          "members": [],
          "name": "getAverage",
          "parameters": [
            {
              "children": [],
              "documentation": "- The first input number",
              "members": [],
              "name": "x",
              "parameters": [],
              "required": false,
              "tags": [
                {
                  "name": "param",
                  "text": "x - The first input number",
                },
              ],
              "type": "number",
            },
            {
              "children": [],
              "documentation": "- The second input number",
              "members": [],
              "name": "y",
              "parameters": [],
              "required": false,
              "tags": [
                {
                  "name": "param",
                  "text": "y - The second input number",
                },
              ],
              "type": "number",
            },
          ],
          "required": false,
          "tags": [
            {
              "name": "remarks",
              "text": "This method is part of the {@link core-library#Statistics Statistics subsystem}.",
            },
            {
              "name": "param",
              "text": "x - The first input number",
            },
            {
              "name": "param",
              "text": "y - The second input number",
            },
            {
              "name": "returns",
              "text": "The arithmetic mean of \`x\` and \`y\`",
            },
            {
              "name": "beta",
              "text": "",
            },
          ],
          "type": "(x: number, y: number) => number",
        },
        {
          "children": [],
          "documentation": "a string variable",
          "members": [],
          "name": "aLongString",
          "parameters": [],
          "required": false,
          "tags": [],
          "type": "string",
        },
        {
          "children": [],
          "documentation": "a property function with x as string parameter",
          "members": [],
          "name": "propertyFn",
          "parameters": [
            {
              "children": [],
              "documentation": "",
              "members": [],
              "name": "x",
              "parameters": [],
              "required": false,
              "tags": [],
              "type": "string",
            },
          ],
          "required": false,
          "tags": [
            {
              "name": "param",
              "text": "x - The first input number",
            },
          ],
          "type": "(x: string) => void",
        },
        {
          "children": [],
          "documentation": "",
          "members": [],
          "name": "aRecursiveTest",
          "parameters": [],
          "required": false,
          "tags": [],
          "type": "Statistics",
        },
      ],
      "documentation": "",
      "members": [],
      "name": "Statistics",
      "parameters": [],
      "required": false,
      "tags": [],
      "type": "any",
    },
  ],
  "documentation": "",
  "members": [],
  "name": "root",
  "parameters": [],
  "required": false,
  "tags": [],
  "type": "",
}
`;
